packages/@lwc/ssr-compiler/src/compile-js/wire.ts

```diff
@@ -0,0 +1,134 @@
+        config?.properties?.map?.((objProp) => {
+            if (!is.property(objProp)) {
+                throw new Error('Object spread syntax is disallowed in @wire config.');
+            }
+            const { key, value } = objProp;
+
+            if (!is.identifier(key)) {
+                throw new Error(`@wire config entry key must be an identifer; found ${key.type}`);
+            }
+            if (!is.literal(value) || typeof value.value !== 'string' || value.value[0] !== '$') {
+                throw new Error('@wire config entry values must be strings starting with a $');
+            }
+            const referencedField = value.value.slice(1);
+            if (!isIdentifierName(referencedField) || isReservedWord(referencedField, true)) {
+                throw new Error(`@wire config referenced invalid field: ${referencedField}`);
+            }
+
+            return {
+                configKey: key.name,
+                referencedField,
+            };
+        }) ?? [];
+
+    return {
+        propName,
+        adapterConstructorId,
+        config: extractedConfig,
+    };
+}
+
+export function catalogWireAdapters(state: ComponentMetaState, node: EsPropertyDefinition) {
+    if (!is.identifier(node.key)) {
+        throw new Error(
+            'Unimplemented: wires that decorate non-identifiers are not currently supported.'
+        );
+    }
+    const { name } = node.key;
+
+    const { decorators } = node as EsPropertyDefinitionWithDecorators;
+    if (decorators?.length !== 1) {
+        throw new Error('Only one decorator can be applied to a single field.');
+    }
+
+    const expression = decorators[0].expression as EsExpression;
+    if (!is.callExpression(expression)) {
+        throw new Error('The @wire decorator must be called.');
+    }
+
+    const { arguments: args } = expression;
+    const [adapterConstructorId, config] = args;
+    if (!is.identifier(adapterConstructorId)) {
+        throw new Error('The @wire decorator must reference a wire adapter class.');
+    }
+    if (config && !is.objectExpression(config)) {
+        throw new Error('Invalid config provided to @wire decorator; expected an object literal.');
+    }
+
+    state.wireAdapters = [
+        ...state.wireAdapters,
+        extractWireConfig(name, adapterConstructorId.name, config),
+    ];
+}
+
+function bWireConfigObj(adapter: WireAdapter) {
+    return b.objectExpression(
+        adapter.config.map(({ configKey, referencedField }) =>
+            b.property(
+                'init',
+                b.identifier(configKey),
+                b.memberExpression(b.identifier('instance'), b.identifier(referencedField))
+            )
+        )
+    );
+}
+
+const bWireAdapterPlumbing = esTemplate`
+    const wireInstance = new ${is.identifier}((newValue) => {
+        instance.${is.identifier} = newValue;
+    });
+    wireInstance.connect?.();
+    if (wireInstance.update) {
+        const wireConfigObj = ${is.objectExpression};
```

Comment:
TODO: this might need to be re-computed inside `__connectContext` callback a few lines below